crimson/onode-staged-tree: fix tree_cursor_t::Cursor to be aware of extent duplication

cyx1231st opened the pull request:
Previously, tree_cursor_t::Cache is not aware of the COWed extent, and still point to the read-only version of value header.
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

cyx1231st commented:
@athanatos 
Now `unittest-seastore` in wip-seastore-2021-03-04-010022 stops at
`ceph/src/crimson/os/seastore/onode_manager/staged-fltree/node_delta_recorder.h:29: ceph::bufferlist crimson::os::seastore::onode::DeltaRecorder::get_delta(): Assertion '!is_empty()' failed.`
Which I think is normal because I see `TODO: fill in recorder` that recorder is not filled yet in `FLTreeOnode` during mutation.

athanatos commented:
I'd be ok with merging https://github.com/ceph/ceph/pull/39828 first and adding a test at the FLTreeOnodeManager level like https://github.com/athanatos/ceph/commit/694fa5ece9c9d205d8565c394ae1082f9e1dac3c .  I think we need functional test coverage at that level, however, as the FLTree implementation is simply too complicated to debug from a higher level.

athanatos commented:
Hmm, https://github.com/athanatos/ceph/commits/sjust/wip-seastore-2021-03-08-124230 with this patch crashes on ./bin/unittest-fltree-onode-manager apparently during cursor destruction (https://gist.github.com/athanatos/cb5582b517f0a34b328320588dc66bb5).

cyx1231st commented:
@athanatos 

I can't reproduce it locally, `unittest-fltree-onode-manager` still stops at `Assertion '!is_empty()'` in my environment, under debug build and `-DWITH_ASAN=ON`.

However, from your gist I can get some clues. Cursor's cached key-view is still pointing to the read-only version of extent even when the extent is duplicated but layout has not changed yet. It should be good normally because `CachedExtent` holds a reference to the `prior_instance`. But it becomes a problem after the transaction submission when try to destruct the `Cursor`, `LeafNode::do_untrack_cursor` will try to validate the key-view, but the `prior_instance` could be already freed.

athanatos commented:
Apologies, that branch didn't actually have the recorder fixes committed.  https://github.com/athanatos/ceph/tree/sjust/wip-seastore-2021-03-08-195742 should do the trick.

cyx1231st commented:
Both `unittest-fltree-onode-manager` and `unittest-seastore` are working now in wip-seastore-2021-03-08-195742, cherry-picking https://github.com/ceph/ceph/pull/39911/commits/c7ab78fad984c1e8f2101747add47ca01e9294c1.

cyx1231st commented:
@athanatos Could you include your commits "fltree_onode_manager: fill in delta recorder" or maybe also "onode_tree: add FLTreeOnodeManager unittest", so I can extend "test_fltree_onode_manager.cc" base on these?

athanatos commented:
@cyx1231st I pushed those two commits as part of https://github.com/ceph/ceph/pull/39828 -- though I commented out the portion of the unit test that currently fails.

cyx1231st commented:
Changeset:
- Cleanup existing test utilities for reuse;
- Implemented FLTreeOnodeManager unittest;
- Fixed a minor issue in FLTreeOnodeManager;

@athanatos please review if the FLTreeOnodeManager unittest is sufficient;

tchaikov reviewed (approved):
LGTM apart from couple nits.

cyx1231st commented on src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h:
```diff
@@ -361,8 +359,8 @@ class TreeBuilder {
       auto iter = kvs.begin();
-      while (!iter.is_end()) {
-        auto p_kv = iter.get_p_kv();
+      while (iter != kvs.end()) {
```
fixed

cyx1231st commented on src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h:
```diff
@@ -371,9 +369,9 @@ class TreeBuilder {
       iter = kvs.begin();
       auto cursor = tree->begin(t).unsafe_get0();
-      while (!iter.is_end()) {
+      while (iter != kvs.end()) {
```
fixed

cyx1231st commented on src/test/crimson/seastore/onode_tree/test_fltree_onode_manager.cc:
```diff
@@ -63,73 +83,147 @@ struct fltree_onode_manager_test_t :
+
+      std::vector<OnodeRef> onodes;
+      onodes.emplace_back(onode);
+      manager->write_dirty(t, onodes).unsafe_get0();
```
fixed
